Fabric.Authorization.Domain/Services/GroupMigratorService.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Fabric.Authorization.Domain.Models;
+using Fabric.Authorization.Domain.Stores;
+using Serilog;
+
+namespace Fabric.Authorization.Domain.Services
+{
+    public class GroupMigratorService
+    {
+        private readonly IGroupStore _groupStore;
+        private readonly ILogger _logger;
+
+        public GroupMigratorService(
+            IGroupStore groupStore,
+            ILogger logger)
+        {
+            _groupStore = groupStore ?? throw new ArgumentNullException(nameof(groupStore));
+            _logger = logger;
+        }
+
+        public async Task<GroupMigrationResult> MigrateDuplicateGroups()
+        {
+            var groups = (await _groupStore.GetAll()).ToList();
+
+            var groupKeys = groups
+                .GroupBy(g => g.Name, StringComparer.OrdinalIgnoreCase)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key);
+
+            var groupMigrationResult = new GroupMigrationResult();
+            foreach (var key in groupKeys)
+            {
+                var duplicateGroups = groups.Where(g => string.Equals(key, g.Name, StringComparison.OrdinalIgnoreCase)).ToList();
+                var originalGroup = duplicateGroups.First();
+                duplicateGroups.RemoveAt(0);
+
+                var groupMigrationRecord = new GroupMigrationRecord
+                {
+                    DuplicateCount = duplicateGroups.Count
+                };
+
+                foreach (var duplicateGroup in duplicateGroups)
+                {
+                    // migrate roles
+                    foreach (var role in duplicateGroup.Roles)
+                    {
+                        var roleExists = originalGroup.Roles.Any(r => r.Equals(role));
```

Comment:
Should we only migrate non deleted Roles?

Reply:
I'll update `GetAll` to filter out all deleted child entities.